src/tools/rbd_mirror/PoolReplayer.cc

```diff
@@ -551,25 +525,228 @@ void PoolReplayer<I>::run()
+          }
+          ctx->complete(r);
+        });
+    namespace_replayer->init(on_init);
+  }
+
+  gather_ctx->activate();
+
+  m_lock.unlock();
+  cond.wait();
+  m_lock.lock();
+
+  C_SaferCond acquire_cond;
+  auto acquire_gather_ctx = new C_Gather(cct, &acquire_cond);
+
+  if (m_leader) {
+    std::vector<std::string> instance_ids;
+    m_leader_watcher->list_instances(&instance_ids);
+
+    for (auto &name : mirroring_namespaces) {
+      auto on_acquire = new FunctionContext(
+          [this, name, &instance_ids, ctx=acquire_gather_ctx->new_sub()](int r) {
+            std::lock_guard locker{m_lock};
+            if (r < 0) {
+              derr << "failed to handle acquire leader for namespace: "
+                   << name << ": " << cpp_strerror(r) << dendl;
+              auto namespace_replayer = m_namespace_replayers[name];
+              m_namespace_replayers.erase(name);
+              auto on_shut_down = new FunctionContext(
+                [this, namespace_replayer, ctx](int r) {
+                  delete namespace_replayer;
+                  ctx->complete(r);
+                });
+              namespace_replayer->shut_down(on_shut_down);
+              return;
+            }
+            m_namespace_replayers[name]->handle_instances_added(instance_ids);
+            ctx->complete(0);
+          });
+      m_namespace_replayers[name]->handle_acquire_leader(on_acquire);
+    }
+    acquire_gather_ctx->activate();
+
+    m_lock.unlock();
+    acquire_cond.wait();
+    m_lock.lock();
+  } else {
+    std::string leader_instance_id;
+    if (m_leader_watcher->get_leader_instance_id(&leader_instance_id)) {
+      for (auto &name : mirroring_namespaces) {
+        m_namespace_replayers[name]->handle_update_leader(leader_instance_id);
+      }
+    }
+  }
+
+  unlock_leader();
 }
 
 template <typename I>
-void PoolReplayer<I>::print_status(Formatter *f, stringstream *ss)
-{
-  dout(20) << "enter" << dendl;
+int PoolReplayer<I>::list_mirroring_namespaces(
+    std::set<std::string> *namespaces) {
+  ceph_assert(ceph_mutex_is_locked(m_lock));
+
+  std::vector<std::string> names;
+
+  int r = librbd::api::Namespace<I>::list(m_local_io_ctx, &names);
+  if (r < 0) {
+    derr << "failed to list namespaces: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  for (auto &name : names) {
+    cls::rbd::MirrorMode mirror_mode = cls::rbd::MIRROR_MODE_DISABLED;
+    int r = librbd::cls_client::mirror_mode_get(&m_local_io_ctx, &mirror_mode);
+    if (r < 0 && r != -ENOENT) {
+      derr << "failed to get namespace mirror mode: " << cpp_strerror(r)
+           << dendl;
+      if (m_namespace_replayers.count(name) == 0) {
+        continue;
+      }
+    } else if (mirror_mode == cls::rbd::MIRROR_MODE_DISABLED) {
+      dout(10) << "mirroring is disabled for namespace " << name << dendl;
+      continue;
+    }
+
+    namespaces->insert(name);
+  }
+
+  return 0;
+}
+
+template <typename I>
+void PoolReplayer<I>::lock_leader() {
```

Comment:
Can you explain this lock/unlock leader stuff? I am not immediately understanding its purpose.

Reply:
It is needed for the case when `PoolReplayer::update_namespace_replayers` detects and adds a new  (or delete non existent) namespace. I don't want that during a namespace replayer initialization or shut down the leader is changed (I can't safely call `handle_acquire|release_leader` for a namespace that is in the process of initialization or shut down). So the `update_namespace_replayers` blocks leader acquire/release callbacks until it updates the namespace replayers.

Reply:
Maybe it would be clearer if it was a "with_namespace_replayers([lambda])" style so you don't need to explicitly lock/unlock in several places -- it would be wrapped by the helper method. Also, what happens if you have multiple leader acquire/release events while it's locked?

Reply:
@dillaman sorry, I missed your question about multiple leader acquire/release events when reading this for the first time. I thought we wouldn't have multiple (simultaneous) events, as they would serialize on `librbd::ManagedLock` layer?